### PR TITLE
Better compiler detection on Linux

### LINKED ISSUE #2733: Move EngineFinder.cmake in new project template to a subfolder
**Is your feature request related to a problem? Please describe.**
Usually the CMake files are located in the `cmake` subfolder. For example: https://github.com/o3de/o3de/tree/development/cmake

**Describe the solution you'd like**
It would be awesome to move `EngineFinder.cmake` to `cmake` (maybe use capital case?) subfolder and include it in the generated `CMakeLists.txt`.

### PR DESCRIPTION
**Pull request by amznestebanpapp**
Also moving EngineFinder.cmake to cmake/ in the templates (Closes #2733)

By default compiler detection searches for other compilers before clang. Specifically, for gcc. Since gcc will be installed in most Linux systems, it ends up being picked over clang.
The search for clang is also not great, it looks for "clang" but most clang versions dont symlink "clang" to the current version (so they are actuallly named something like "clang-12/clang++-12").
This change detects the version of clang installed and sets it as `CMAKE_${lang}_COMPILER` so its picked by CMake's compiler detection.
Since compiler detection happens during the call to `project(`, I had to add this to the template and to the engine's root `CMakeLists.txt`

The change in the project is not needed to be propagated to current projects. Its an improvement if projects take the new template.

Since I was adding this to the template, I also tackled issue #2733

### CODE REVIEW

File: cmake/Platform/Linux/CompilerSettings.cmake [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#
+# Copyright (c) Contributors to the Open 3D Engine Project.
+# For complete copyright and license terms please see the LICENSE at the root of this distribution.
+#
+# SPDX-License-Identifier: Apache-2.0 OR MIT
+#
+#
+
+if(NOT CMAKE_C_COMPILER AND NOT CMAKE_CXX_COMPILER)
```
Comment (spham-amzn) [inline cmake/Platform/Linux/CompilerSettings.cmake:9]: This should also check against the environment variables:
`CC` and `CXX` whether it is setup or not. CMake will key from them if CMAKE_C_COMPILER and CMAKE_CXX_COMPILER is not set.

https://cmake.org/cmake/help/latest/envvar/CC.html
https://cmake.org/cmake/help/latest/envvar/CXX.html

File: Templates/DefaultProject/Template/cmake/Platform/Linux/CompilerSettings.cmake [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+if(NOT CMAKE_C_COMPILER AND NOT CMAKE_CXX_COMPILER)
+    set(path_search
+        /bin
+        /usr/bin
+        /usr/local/bin
+        /sbin
+        /usr/sbin
+        /usr/local/sbin
+    )
+    list(TRANSFORM path_search APPEND "/clang-[0-9]*")
```
Comment (lumberyard-employee-dm) [inline Templates/DefaultProject/Template/cmake/Platform/Linux/CompilerSettings.cmake:18]: Technically this regex could be `/clang-[1-9]+`

Comment (amzn-phist) [reply Templates/DefaultProject/Template/cmake/Platform/Linux/CompilerSettings.cmake:18]: What about `clang-10` though?  Maybe `clang-[0-9]+` instead.

**Review by amzn-phist — APPROVED** [review]

File: Templates/DefaultProject/Template/cmake/Platform/Linux/CompilerSettings.cmake (hunk 2 above)
Comment (lumberyard-employee-dm) [reply Templates/DefaultProject/Template/cmake/Platform/Linux/CompilerSettings.cmake:18]: Well `/clang-[1-9]+` still matches "**clang-1** in "clang-10"

Comment (amznestebanpapp) [reply Templates/DefaultProject/Template/cmake/Platform/Linux/CompilerSettings.cmake:18]: should I make the "-" optional?
it also looks like clang 9 was `clang-9.0`, so I have to add the dot there

Comment (amznestebanpapp) [reply Templates/DefaultProject/Template/cmake/Platform/Linux/CompilerSettings.cmake:18]: actually, this is not feeding a regex but a glob search, I am just trying to differentiate it from `clang-cpp-12`

**Review by spham-amzn — APPROVED** [review]

**Review by lumberyard-employee-dm — APPROVED** [review]

**Pull request merged**